scipy/sparse/linalg/_isolve/tests/test_iterative.py

```diff
@@ -507,55 +509,49 @@ def test_x0_working(solver):
         kw = dict(atol=0, tol=1e-6)
 
     x, info = solver(A, b, **kw)
-    assert_equal(info, 0)
-    assert_(np.linalg.norm(A.dot(x) - b) <= 1e-6*np.linalg.norm(b))
+    assert info == 0
+    assert np.linalg.norm(A @ x - b) <= 1e-6 * np.linalg.norm(b)
 
     x, info = solver(A, b, x0=x0, **kw)
-    assert_equal(info, 0)
-    assert_(np.linalg.norm(A.dot(x) - b) <= 1e-6*np.linalg.norm(b))
+    assert info == 0
+    assert np.linalg.norm(A @ x - b) <= 2e-6 * np.linalg.norm(b)
 
 
-@pytest.mark.parametrize('solver', [cg, cgs, bicg, bicgstab, gmres, qmr,
-                                    minres, lgmres, gcrotmk])
-def test_x0_equals_Mb(solver):
-    for case in params.cases:
-        if solver in case.skip:
-            continue
-        with suppress_warnings() as sup:
-            sup.filter(DeprecationWarning, ".*called without specifying.*")
-            A = case.A
-            b = case.b
-            x0 = 'Mb'
-            tol = 1e-8
-            x, info = solver(A, b, x0=x0, tol=tol)
+def test_x0_equals_Mb(solver, case):
+    if solver in case.skip or solver is tfqmr:
+        pytest.skip("unsupported combination")
+    with suppress_warnings() as sup:
+        sup.filter(DeprecationWarning, ".*called without specifying.*")
+        A = case.A
+        b = case.b
+        x0 = 'Mb'
+        tol = 1e-8
+        x, info = solver(A, b, x0=x0, tol=tol)
 
-            assert_array_equal(x0, 'Mb')  # ensure that x0 is not overwritten
-            assert_equal(info, 0)
-            assert_normclose(A.dot(x), b, tol=tol)
+        assert_array_equal(x0, 'Mb')  # ensure that x0 is not overwritten
+        assert info == 0
+        assert_normclose(A @ x, b, tol=tol)
 
 
-@pytest.mark.parametrize(('solver', 'solverstring'), [(tfqmr, 'TFQMR')])
-def test_show(solver, solverstring, capsys):
+def test_show(case, capsys):
     def cb(x):
-        count[0] += 1
+        pass
 
-    for i in [0, 20]:
-        case = params.cases[i]
-        A = case.A
-        b = case.b
-        count = [0]
-        x, info = solver(A, b, callback=cb, show=True)
-        out, err = capsys.readouterr()
-        if i == 20:  # Asymmetric and Positive Definite
-            assert_equal(out, f"{solverstring}: Linear solve not converged "
-                              f"due to reach MAXIT iterations {count[0]}\n")
-        else:  # 1-D Poisson equations
-            assert_equal(out, f"{solverstring}: Linear solve converged due to "
-                              f"reach TOL iterations {count[0]}\n")
-        assert_equal(err, '')
```

Comment:
If I read this test correctly, it iterates through the cases and has a very specific error expectation for exactly the 20th Case. I counted and hope that that's actually `"rand-cmplx-sym-pd-F"`. I'm also not sure about the necessity of the callback here at all (and how the counter through the cases has anything to do with the iterations), but for now I thought I'd keep passing one.

Reply:
Aside from counting incorrectly (off by one), the analysis was correct. :)